iceoryx_examples/automotive_soa/include/owl/kom/event_subscriber.inl

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_SUBSCRIBER_INL
+#define IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_SUBSCRIBER_INL
+
+#include "owl/kom/event_subscriber.hpp"
+
+namespace owl
+{
+namespace kom
+{
+template <typename T>
+inline EventSubscriber<T>::EventSubscriber(const core::String& service,
+                                           const core::String& instance,
+                                           const core::String& event) noexcept
+    : m_subscriber({service, instance, event},
+                   {QUEUE_CAPACITY, HISTORY_REQUEST, iox::NodeName_t(), NOT_OFFERED_ON_CREATE})
+{
+}
+
+template <typename T>
+inline void EventSubscriber<T>::Subscribe(std::size_t) noexcept
+{
+    // maxSampleCount is ignored, because it is an argument to the c'tor of m_subscriber as part of SubscriberOptions.
+    // Utilizing late initalization by wrapping m_subscriber in an cxx::optional and calling the c'tor here would be an
+    // option
+    m_subscriber.subscribe();
+}
+
+template <typename T>
+inline void EventSubscriber<T>::Unsubscribe() noexcept
+{
+    m_subscriber.unsubscribe();
+}
+
+template <typename T>
+template <typename Callable>
+inline core::Result<size_t> EventSubscriber<T>::GetNewSamples(Callable&& callable, size_t maxNumberOfSamples) noexcept
+{
+    IOX_DISCARD_RESULT(maxNumberOfSamples);
+
+    core::Result<size_t> numberOfSamples{0};
+
+    while (m_subscriber.take()
+               .and_then([&](const auto& sample) {
+                   callable(sample.get());
```

Comment:
Since the `Callable` type allows callables of type `cxx::function`, `cxx::function_ref` or `std::function` which are all nullable. How do you make sure that `callable` is actually set?
You also allow here lambda expressions which cannot be null, so you end up here in some kind of a pickle.

I would suggest to get rid of the `Callable` template parameter and switch to `cxx::function`.

Reply:
Well, unfortunately that's the API, hence it's not possible to get rid off it. I added a early return in case the `Callable` is nulled.

Reply:
Then I suggest adjusting the API to a `cxx::function` and introduce the `is_lambda` trait in a follow up PR and extend with this PR this API again.

Reply:
I'm not in favour of an changing API from the spec on this PR. I added a todo item to #1332 and will implement it in a follow-up as we already have a comment regarding this issue in the code. Is that ok for you?

Reply:
Again, I get more and more the feeling this does not belong into iceoryx. In my opinion API from outside is not enforced into iceoryx.
In this case the API introduces problems which we address in iceoryx with verious idioms and techniques and now we should introduce bugs in an example just because an even unnamed API (oh we are not allowed to say autosar and ara com) enforces this.

Why not create another repository like `ara_on_iceoryx` and use iceoryx as dependency?

Reply:
Again, I know what you mean. However, as I wrote previously I'm strongly in favour of keep all the examples, docs and code in one repository.

Reply:
Since we do not have to stick strictly to the automotive API I would still recommend to use `cxx::function` here as `callable` type. Just to make others aware that this generic approach comes with problems.

Reply:
I'm not sure here, as I would prefer not to deviate too much. For me the comment with the hint is ok.